python-package/lightgbm/basic.py

```diff
@@ -323,19 +325,22 @@ def _json_default_with_numpy(obj: Any) -> Any:
         return obj
 
 
+def _to_string(x: Union[int, float, str, List]) -> str:
+    if isinstance(x, list):
+        val_list = ",".join(str(val) for val in x)
+        return f"[{val_list}]"
+    else:
+        return str(x)
+
+
 def _param_dict_to_str(data: Optional[Dict[str, Any]]) -> str:
     """Convert Python dictionary to string, which is passed to C API."""
     if data is None or not data:
         return ""
     pairs = []
     for key, val in data.items():
         if isinstance(val, (list, tuple, set)) or _is_numpy_1d_array(val):
-            def to_string(x):
```

Comment:
this is a good example of what this PR hopes to avoid...this `to_string()` method was being redefined for every element in `data` passed to every call to `_param_dict_to_str()`.